src/lib.rs

```diff
@@ -547,6 +714,131 @@ impl Sender {
 
         Ok(true)
     }
+
+    /// Send the specified message, waiting for sufficient contiguous space to become available in the ring buffer
+    /// if necessary. Uses the lower level `ShmWriter` trait to write messages directly into the mmap buffer.
+    ///
+    /// The size of the message must be greater than zero or else this method will return
+    /// `Err(`[`Error::ZeroSizedMessage`](enum.Error.html#variant.ZeroSizedMessage)`))`.  If the message size is
+    /// greater than the ring buffer capacity, this method will return
+    /// `Err(`[`Error::MessageTooLarge`](enum.Error.html#variant.MessageTooLarge)`))`.
+    pub fn send_direct(&self, value: &impl ShmWriter) -> Result<()> {
+        self.send_direct_timeout_0(value, false, None).map(drop)
+    }
+
+    /// Send the specified message, waiting for sufficient contiguous space to become available in the ring buffer
+    /// if necessary, but only up to the specified timeout. Uses the lower level `ShmWriter` trait to write messages 
+    /// directly into the mmap buffer.
+    /// 
+    /// This will return `Ok(true)` if the message was sent, or `Ok(false)` if it timed out while waiting.
+    ///
+    /// The size of the message must be greater than zero or else this method will return
+    /// `Err(`[`Error::ZeroSizedMessage`](enum.Error.html#variant.ZeroSizedMessage)`))`.  If the size is
+    /// greater than the ring buffer capacity, this method will return
+    /// `Err(`[`Error::MessageTooLarge`](enum.Error.html#variant.MessageTooLarge)`))`.
+    pub fn send_direct_timeout(&self, value: &impl ShmWriter, timeout: Duration) -> Result<bool> {
+        self.send_direct_timeout_0(value, false, Some(timeout))
+    }
+
+    /// Send the specified message, waiting for the ring buffer to become completely empty first. Uses the lower 
+    /// level `ShmWriter` trait to write messages directly into the mmap buffer.
+    /// 
+    /// This method is appropriate for sending time-sensitive messages where buffering would introduce undesirable
+    /// latency.
+    ///
+    /// The size of the message must be greater than zero or else this method will return
+    /// `Err(`[`Error::ZeroSizedMessage`](enum.Error.html#variant.ZeroSizedMessage)`))`.  If the size
+    /// is greater than the ring buffer capacity, this method will return
+    /// `Err(`[`Error::MessageTooLarge`](enum.Error.html#variant.MessageTooLarge)`))`.
+    pub fn send_direct_when_empty(&self, value: &impl ShmWriter) -> Result<()> {
+        self.send_direct_timeout_0(value, true, None).map(drop)
+    }
+
+    fn send_direct_timeout_0(
```

Comment:
Perhaps we could reduce code duplication here by re-implementing `send_timeout_0` in terms of `send_direct_timeout_0`, using a `bincode`-based `ShmWriter` implementation?